### PR TITLE
ENH: Allow `ctypeslib.load_library` to take any path-like object

### PR DESCRIPTION
**Pull request by BvB93**
This PR allows `ctypeslib.load_library()` take take any [path-like](https://docs.python.org/3/glossary.html#term-path-like-object) object,
which includes the likes of strings, bytes and any object implementing the [`__fspath__`](https://docs.python.org/3/library/os.html#os.PathLike) protocol.

### CODE REVIEW

File: numpy/tests/test_ctypeslib.py [hunk 1]
```diff
@@ -37,13 +38,15 @@
                     reason="Known to fail on cygwin")
 class TestLoadLibrary:
     def test_basic(self):
-        try:
```
Comment (BvB93) [inline numpy/tests/test_ctypeslib.py:40]: Note that the `try` / `except` clause is redundant due to the pytest fixture:

https://github.com/numpy/numpy/blob/a9edc0433e2efff6695c2d22a1cd0cad325556ae/numpy/tests/test_ctypeslib.py#L35-L39

File: numpy/ctypeslib.py [hunk 2]
```diff
@@ -119,6 +119,10 @@ def load_library(libname, loader_path):
             warnings.warn("All features of ctypes interface may not work "
                           "with ctypes < 1.0.1", stacklevel=2)
 
+        # Convert path-like objects into strings
+        libname = os.fsdecode(libname)
+        loader_path = os.fsdecode(loader_path)
```
Comment (eric-wieser) [inline numpy/ctypeslib.py:124]: Shouldn't this be `fspath`, or am I missing something?

Comment (eric-wieser) [reply numpy/ctypeslib.py:124]: Just reread the pep - this seems fine too

File: numpy/ctypeslib.py [hunk 3]
```diff
@@ -97,10 +97,10 @@ def load_library(libname, loader_path):
 
         Parameters
         ----------
-        libname : str
+        libname : path-like
             Name of the library, which can have 'lib' as a prefix,
             but without an extension.
-        loader_path : str
+        loader_path : path-like
             Where the library can be found.
```
Comment (eric-wieser) [inline numpy/ctypeslib.py:104]: Can you add a`..versionchanged` somewhere?

File: numpy/ctypeslib.py [hunk 4]
```diff
@@ -97,10 +97,10 @@ def load_library(libname, loader_path):
 
         Parameters
         ----------
-        libname : str
+        libname : path-like
```
Comment (eric-wieser) [inline numpy/ctypeslib.py:100]: Are paths with / actually allowed here? If not, this looks like it should remain a str

Comment (BvB93) [reply numpy/ctypeslib.py:100]: It seems to be, yes.
What happens when defining `libpath` is, roughly speaking, calling  `os.path.join` on `loader_path` and `libname` the input parameters.

https://github.com/numpy/numpy/blob/4afd82d8dd2c1ef48c62bdf875a194e407f2d2d3/numpy/ctypeslib.py#L144-L148

File: numpy/ctypeslib.py (hunk 3 above)
Comment (BvB93) [reply numpy/ctypeslib.py:104]: Will this need a release note as well?
I figure the change was small enough, but it can't hurt double checking.

Comment (eric-wieser) [reply numpy/ctypeslib.py:104]: I'd probably write a short one, yes. No need to hit the mailing list.

File: numpy/ctypeslib.py (hunk 4 above)
Comment (eric-wieser) [reply numpy/ctypeslib.py:100]: Hmm, to me that implies the path component should always be in `loader_path`, and `libname` should always be a string.

File: numpy/ctypeslib.py (hunk 3 above)
Comment (BvB93) [reply numpy/ctypeslib.py:109]: Let's do it; `versionchanged` was added in https://github.com/numpy/numpy/pull/17530/commits/3271a08baf9500f0112be24e7573d9ae2945f847 and the release note in https://github.com/numpy/numpy/pull/17530/commits/d690cf60beef0739f6d7c15e778c37e528afe58f.

File: numpy/ctypeslib.py (hunk 4 above)
Comment (BvB93) [reply numpy/ctypeslib.py:105]: Perhaps, though `os.path.join` is not required to join a path and a file name; 
the joining of two paths is also perfectly acceptable.

``` python
In [1]: import os

In [2]: os.path.join('a/b/c/', 'd/e/f')
Out[2]: 'a/b/c/d/e/f'
```

Besides, why should it remain a string if a / is not allowed?
The likes of `pathlib.Path` are just generally useful objects, regardless of whether it is used to represent a path- or file-name.

**Comment by charris** [comment]
Needs a release note.

**Comment by BvB93** [comment]
> Needs a release note.

See https://github.com/numpy/numpy/pull/17530/commits/d690cf60beef0739f6d7c15e778c37e528afe58f.

**Comment by charris** [comment]
@eric-wieser Look good to you?

**Comment by charris** [comment]
close/reopen

**Comment by charris** [comment]
@eric-wieser Ping.

**Comment by charris** [comment]
Thanks Bas.

**Pull request merged**